### PR TITLE
Pending service update

### LINKED ISSUE #1225: PendingDocServic - Update
- Add Update to Service. Delegate the model Derivation to documents.Service Derive
- Store the model and return the model

### PR DESCRIPTION
**Pull request by mikiquantum**
closes #1225

### CODE REVIEW

File: documents/invoice/model.go [hunk 1]
```diff
@@ -624,6 +624,15 @@ func (i *Invoice) unpackFromUpdatePayload(payload documents.UpdatePayload) (*Inv
 	}, nil
 }
 
+// Patch merges payload data into model
+func (i *Invoice) Patch(payload documents.UpdatePayload) (documents.Model, error) {
+	inv, err := i.unpackFromUpdatePayload(payload)
```
Comment (vedhavyas) [inline documents/invoice/model.go]: I see a problem here. `unpackFromUpdatePayload` internally calls `PreparesNewVersion` making it skip the current version without anchoring it.

I think we need to split the unpackFromUpdatePayload to patching and preparing new version. WDYT ?

File: documents/service.go [hunk 2]
```diff
@@ -28,6 +28,12 @@ type DocumentProof struct {
 	FieldProofs []*proofspb.Proof
 }
 
+// Patcher interface defines a Patch method for inner Models
+type Patcher interface {
+	// Patch merges payload data into model
+	Patch(payload UpdatePayload) (Model, error)
```
Comment (vedhavyas) [inline documents/service.go]: Since we anyway call this on the model, we dont need to return the model I think

File: pending/service.go [hunk 3]
```diff
@@ -65,6 +71,35 @@ func (s service) Create(ctx context.Context, payload documents.UpdatePayload) (d
...
+	if err != nil {
+		return nil, contextutil.ErrDIDMissingFromContext
+	}
+
+	m, err := s.pendingRepo.Get(accID[:], payload.DocumentID)
+	if err != nil {
+		return nil, err
+	}
+
+	if m.GetStatus() != documents.Pending {
```
Comment (vedhavyas) [inline pending/service.go]: this check is will always fail, no ?
since the pending repo will always store the pending documents, WDYT ?

Comment (mikiquantum) [reply pending/service.go]: right, since the create will take care of the already anchored flow.

File: documents/invoice/model.go (hunk 1 above)
Comment (mikiquantum) [reply documents/invoice/model.go]: you are right, didn't catch that, I will separate those now.

File: documents/coredocument.go [hunk 4]
```diff
@@ -241,6 +241,41 @@ func (cd *CoreDocument) AppendSignatures(signs ...*coredocumentpb.Signature) {
 	cd.Modified = true
 }
 
+// PrepareNewData overrides only core document data without provisioning new versions since for document updates
+func (cd *CoreDocument) PrepareNewData(documentPrefix []byte, collaborators CollaboratorsAccess, attrs map[AttrKey]Attribute) (*CoreDocument, error) {
```
Comment (vedhavyas) [inline documents/coredocument.go]: I think we should check for status==pending only. WDYT ?

File: documents/error.go [hunk 5]
```diff
@@ -50,6 +50,9 @@ const (
 	// ErrDocumentProof must be used when document proof creation fails
 	ErrDocumentProof = errors.Error("document proof error")
 
+	// ErrNotPatcher must be used if an expected patcher model is not a patcher
+	ErrNotPatcher = errors.Error("model not patcher")
```
Comment (vedhavyas) [inline documents/error.go]:
```suggestion
	ErrNotPatcher = errors.Error("document doesn't support patching")
```

File: pending/service.go [hunk 6]
```diff
@@ -15,9 +15,15 @@ var srvLog = logging.Logger("pending-service")
 // ErrPendingDocumentExists is a sentinel error used when document was created and tried to create a new one.
 const ErrPendingDocumentExists = errors.Error("Pending document already created")
 
+// ErrInProcessDocument is a sentinel error used when a document is not in pending state
+const ErrInProcessDocument = errors.Error("document is not in pending state")
```
Comment (vedhavyas) [inline pending/service.go]: Forgot to use this in the CoreDocument PrepareNewData ?

**Pull request merged**